e2e/nomostest/clean.go

```diff
@@ -372,198 +313,207 @@ func resetSystemNamespaces(nt *NT, failOnError FailOnError) {
 		}
 		if differ.IsManageableSystemNamespace(&ns) {
 			if err := deleteConfigSyncAndTestAnnotationsAndLabels(nt, &ns); err != nil {
-				nt.T.Log(err)
-				errDeleting = true
+				return err
 			}
 			if err := nt.Validate(ns.Name, "", &corev1.Namespace{}, namespaceHasNoConfigSyncAnnotationsAndLabels()); err != nil {
-				if failOnError {
-					nt.T.Fatal(err)
-				} else {
-					nt.T.Log(err)
-				}
-			}
-		}
-	}
-	if errDeleting && bool(failOnError) {
-		nt.T.Fatal("error resetting the system namespaces")
-	}
-}
-
-// deleteImplicitNamespaces deletes the namespaces with the PreventDeletion annotation.
-func deleteImplicitNamespaces(nt *NT, failOnError FailOnError) {
-	errDeleting := false
-	nsList := &corev1.NamespaceList{}
-	if err := nt.Client.List(nt.Context, nsList); err != nil {
-		nt.T.Log(err)
-		errDeleting = true
-	}
-	for _, ns := range nsList.Items {
-		if annotation, ok := ns.Annotations[common.LifecycleDeleteAnnotation]; ok && annotation == common.PreventDeletion {
-			if err := nt.Client.Delete(nt.Context, &ns); err != nil {
-				nt.T.Log(err)
-				errDeleting = true
-			}
-			if err := WatchForNotFound(nt, kinds.Namespace(), ns.Name, ""); err != nil {
-				if failOnError {
-					nt.T.Fatal(err)
-				} else {
-					nt.T.Error(err)
-				}
+				return err
 			}
 		}
 	}
-	if errDeleting && bool(failOnError) {
-		nt.T.Fatal("error deleting implicit namespaces")
-	}
+	return nil
 }
 
-// removeKubevirt cleans up the Kubevirt resources when the kubevirt namespace is stuck
-// in the `Terminating` phase.
-func removeKubevirt(nt *NT, failOnError FailOnError) {
+// removeKubevirt cleans up the Kubevirt resources when the kubevirt namespace
+// is stuck in the `Terminating` phase.
+func removeKubevirt(nt *NT) error {
+	nt.T.Log("[CLEANUP] Cleaning up kubevirt resources")
+	start := time.Now()
 	defer func() {
-		nt.MustKubectl("delete", "crd", "-l", "app.kubernetes.io/component=kubevirt")
+		elapsed := time.Since(start)
+		nt.T.Logf("[CLEANUP] Cleaning up kubevirt resources took %v", elapsed)
 	}()
+
 	kubevirtNS := corev1.Namespace{}
 	if err := nt.Get("kubevirt", "", &kubevirtNS); err != nil {
-		if apierrors.IsNotFound(err) {
-			return
+		if !apierrors.IsNotFound(err) {
+			return err
 		}
-		if failOnError {
-			nt.T.Fatalf("failed to get the Kubevirt namespace", err)
-		} else {
-			nt.T.Logf("failed to get the Kubevirt namespace", err)
+	} else {
+		// Remove finalizers on KubeVirt object to skip cleanup, if it exists.
+		// WARNING: This may leave behind GCP resources on GKE clusters that
+		// support nested virtualization, like those with N2 nodes.
+		// TODO: Find a safer way to clean up KubeVirt...
```

Comment:
Maybe a better solution would be to use a different resource type in our test case 🤷

Reply:
All I did was refactor what we have here and add a warning. If you want to take a stab at not using KubeVirt in our tests, be my guest.

Reply:
I wasn't saying you need to do this for this PR, I still see your change as an improvement. I was just being a bit conversational with your TODO